<a href="https://colab.research.google.com/github/kitawa333-svg/air-quality-eda/blob/main/Indian_AQI_Data_Processing_PRAC_1_ass.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#  **Indian Air Quality Data Analysis**

---


We are analyzing air quality data from **26 Indian cities** (2015–2019) 📊.  
The goal is to **combine city datasets**, **explore pollution patterns**, and **build predictive models** to understand and forecast air quality trends.

---



# **1. Dataset Rationale and Project Objective**

Air pollution is a **major public health crisis** in Indian cities, with particulate matter (PM2.5, PM10) and gaseous pollutants (NO2, CO, SO2, O3) contributing to respiratory diseases, cardiovascular issues, and premature deaths.

**Why this dataset?**  
The Indian Air Quality dataset provides **daily measurements** across multiple pollutants along with the **Air Quality Index (AQI)** — a standardized health-risk indicator. Analyzing pollutant-AQI relationships identifies **which contaminants drive poor air quality**, informing targeted intervention strategies.

### **Main Objectives:**

1. **Exploratory Data Analysis (EDA)**  
   - Understand pollution patterns, trends, and spatial distributions  
   - Identify relationships between pollutants and AQI  
   - Analyze seasonal and city-level variations

2. **Predictive Modeling**  
   - Build a regression model to estimate AQI from pollutant concentrations  
   - Enable basic air quality forecasting for early warnings

**Real-world application:** Predictive AQI models support **public health advisories**, **pollution control policies**, and **individual precautionary measures** during high-pollution episodes.

---



In [ ]:
#libraries for data analysis
import pandas as pd
import math
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import os

### **🔧 Tools for Analysis**

**Pandas & NumPy**: For data manipulation and numerical computations  
**Matplotlib & Seaborn**: For static visualizations and statistical plots  
**Plotly**: For interactive charts (if used in final presentation)  
**OS & Glob**: For file system operations and batch file loading

---



### **Mount the drive**

---



In [ ]:
# Mount Google Drive - run ONCE at start
try:
    from google.colab import drive
    drive.mount('/content/drive')
    print("✓ Drive mounted successfully")

    # List your folder
    drive_path = '/content/drive/MyDrive/Indian air quality'
    !ls "{drive_path}"

except Exception as e:
    print(f"Mount failed: {e}")
    print("Please restart runtime and try again")

Mounted at /content/drive
✓ Drive mounted successfully
Ahmedabad_data.csv	 Coimbatore_data.csv  Lucknow_data.csv
Aizawl_data.csv		 Delhi_data.csv       model.pkl
Amaravati_data.csv	 Ernakulam_data.csv   Mumbai_data.csv
Amritsar_data.csv	 Gurugram_data.csv    Patna_data.csv
Bengaluru_data.csv	 Guwahati_data.csv    Shillong_data.csv
Bhopal_data.csv		 Hyderabad_data.csv   Talcher_data.csv
Brajrajnagar_data.csv	 Jaipur_data.csv      Thiruvananthapuram_data.csv
Chandigarh_data.csv	 Jorapokhar_data.csv  Visakhapatnam_data.csv
Chennai_data.csv	 Kochi_data.csv
cleaned_air_quality.csv  Kolkata_data.csv


### **📁 Data Access Setup**

Google Drive is mounted to access the dataset stored in the cloud. The folder `Indian air quality` contains individual CSV files for each city — a common format for multi-location environmental data.

---



### **📂 Available Data Files**

The directory contains 15+ city-specific CSV files (e.g., `Ahmedabad_data.csv`, `Delhi_data.csv`), each following the naming convention `{City}_data.csv`. This structured naming allows automated loading of all files.

---



# **2. Combain CSV files into one CSV file**

---



In [ ]:
import glob #find files by name patterns

**OS & Glob**: For file system operations and batch file loading

In [ ]:
city_files = glob.glob(os.path.join(drive_path, "*_data.csv")) #find files that have'.csv'

all_cities_data = [] #putting each city's data in this list b4 combining them

for file_name in city_files:

    city_df = pd.read_csv(file_name) #read the current city's CSV file into a DataFrame

    all_cities_data.append(city_df) #adds the cityies data to our list

    print(f"Loaded: {file_name}") #print which file we just read

combined_data = pd.concat(all_cities_data, ignore_index=True) #pd.concat() joins all the DataFrames in our list together, ignore_index=True - makes row numbers continuous

combined_data.to_csv("all_cities_combined.csv", index=False) #index=False - don't save the row numbers as a separate column

print(f"SUCCESS: Combined {len(city_files)} city files into one file with {len(combined_data)} total rows") #len(city_files) = count of how many city files are combined; len(combined_data) = total no. of rows in the final file
print("The combined file is saved as: all_Indian_cities_combined.csv")

Loaded: /content/drive/MyDrive/Indian air quality/Guwahati_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Patna_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Kochi_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Hyderabad_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Visakhapatnam_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Ahmedabad_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Jorapokhar_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Kolkata_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Delhi_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Ernakulam_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Shillong_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Amaravati_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Bhopal_data.csv
Loaded: /content/drive/MyDrive/Indian air quality/Coimbatore_data.csv
Loaded: /content/drive/MyDrive/Indian air quali

### **🔄 Data Integration Process**

**Step 1: File Discovery** – Using `glob` to find all `*_data.csv` files  
**Step 2: Sequential Loading** – Reading each city's data into memory  
**Step 3: Consolidation** – Combining all DataFrames vertically (row-wise)  
**Step 4: Persistence** – Saving the unified dataset for repeated use

**Rationale**: Combining city files enables cross-city analysis and creates a single source for all subsequent operations.

---



### **✅ Data Integration Result**

Successfully merged city-specific datasets into a single DataFrame containing **29,531 daily measurements** from **26 cities** over **5 years**. The unified structure enables comprehensive pan-India air quality analysis.

---



**Load and Inspect Combined Dataset**

---



In [ ]:
df= pd.read_csv('all_cities_combined.csv') #reads the csv files into the Dataframe
print (df) #prints the contents of the new csv file

            City        Date   PM2.5    PM10    NO    NO2    NOx    NH3    CO  \
0       Guwahati  16/02/2019  124.77  187.08   NaN    NaN    NaN    NaN   NaN   
1       Guwahati  17/02/2019   50.82  118.01  4.03   9.30  16.56  29.24  0.48   
2       Guwahati  18/02/2019   58.31   85.39  3.67  10.82  17.63  23.60  0.47   
3       Guwahati  19/02/2019   59.88   78.04  3.16   9.65  15.51  19.03  0.51   
4       Guwahati  20/02/2019   55.61   92.36  2.89  11.71  17.30  16.68  0.49   
...          ...         ...     ...     ...   ...    ...    ...    ...   ...   
29526  Bengaluru  27/06/2020   16.60   29.48  3.06  13.68  13.07   6.88  0.67   
29527  Bengaluru  28/06/2020   20.44   26.34  2.69  10.33  10.58   6.58  0.66   
29528  Bengaluru  29/06/2020   28.68   29.27  3.62  12.12  12.94   6.80  0.56   
29529  Bengaluru  30/06/2020   14.47   24.26  4.61  12.69  15.00   6.82  0.56   
29530  Bengaluru  01/07/2020   17.50   30.48  3.95  13.25  14.83   7.42  0.54   

         SO2     O3  Benzen

In [ ]:
#data set details
print(f" YOUR Dataset Size: {df.shape[0]} rows × {df.shape[1]} columns")
print(f" Date Range: {df['Date'].min()} to {df['Date'].max()}")
print(f"Cities: {df['City'].nunique()} unique cities")

 YOUR Dataset Size: 29531 rows × 16 columns
 Date Range: 01/01/2015 to 31/12/2019
Cities: 26 unique cities


### **🔍 Data Overview**

Dataset loaded successfully with **16 columns** including pollutants (PM2.5, PM10, NO2, CO, SO2, O3), location (City, Date), and AQI. The **29,531 daily records** span from **01/01/2015 to 31/12/2019** across **26 Indian cities**, providing comprehensive air quality coverage.

---



In [ ]:
df.shape #show the number of columns and rows
print(f'No of Rows: {df.shape[0]}, No of Columns: {df.shape[1]}') #shows the no. of rows and columns

No of Rows: 29531, No of Columns: 16


### **📏 Dataset Size**

**29,531 rows × 16 columns** — This represents daily air quality measurements collected over 5 years across 26 major Indian cities. The volume allows for robust temporal and spatial trend analysis.

---



In [ ]:
df.head() #shows the first 5 rows of the data set

City        Date   PM2.5    PM10    NO    NO2    NOx    NH3    CO  \
0  Guwahati  16/02/2019  124.77  187.08   NaN    NaN    NaN    NaN   NaN   
1  Guwahati  17/02/2019   50.82  118.01  4.03   9.30  16.56  29.24  0.48   
2  Guwahati  18/02/2019   58.31   85.39  3.67  10.82  17.63  23.60  0.47   
3  Guwahati  19/02/2019   59.88   78.04  3.16   9.65  15.51  19.03  0.51   
4  Guwahati  20/02/2019   55.61   92.36  2.89  11.71  17.30  16.68  0.49   

     SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
0    NaN    NaN      NaN      NaN     NaN    NaN           NaN  
1  14.03  17.67     1.07      NaN     NaN   92.0  Satisfactory  
2  14.47  33.85     0.88      NaN     NaN  122.0      Moderate  
3  16.42  41.23     1.00      NaN     NaN   80.0  Satisfactory  
4  27.07  44.89     0.98      NaN     NaN  105.0      Moderate

### **👀 First Glimpse**

The first 5 rows show sample readings with pollutant concentrations, dates, and city names. Notice the AQI values already range widely, indicating diverse air quality conditions across locations and seasons.

---



In [ ]:
df.columns #lists all column names

Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

In [ ]:
df.info() #summary of data types, column names and null counts

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


### **📋 Column Information**  

**Temporal & Spatial:**
- Date: Daily readings (2015-2020)
- City: 10+ Indian cities including Delhi, Mumbai, Ahmedabad

**Primary Pollutants (µg/m³):**
- PM2.5, PM10: Particulate matter
- NO, NO2, NOx: Nitrogen oxides  
- CO, SO2, O3: Gaseous pollutants

**Volatile Organic Compounds (VOCs):**
- Benzene, Toluene, Xylene

**Air Quality Indicators:**
- AQI: Numerical index (0-500)
- AQI_Bucket: Categorical severity (Good, Moderate, Unhealthy, etc.)

**Data Types:** All numeric columns are float64, suitable for analysis without conversion.

---



### **📋 Data Completeness Assessment**

The dataset contains **29,531 daily observations** across **16 columns**, recording air quality measurements from **26 Indian cities (2015–2019)**.

**Critical variables** (AQI, City, Date) have complete data — essential for analysis. However, **pollutant columns show significant gaps**:

- **Severe gaps (>30%)**: Xylene (61%), PM10 (38%), NH3 (35%)
- **Moderate gaps (10-20%)**: PM2.5 (16%), AQI_Bucket (16%), NO2 (12%)
- **Minor gaps (<10%)**: CO (7%)

These gaps reflect **real-world monitoring limitations** rather than data quality issues.

---



In [ ]:
# Function to calculate missing values percentage
def missing_values_table(df):
    mis_val = df.isnull().sum()
    mis_val_percent = 100 * mis_val / len(df)
    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)
    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})
    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)
    return mis_val_table

# Display missing values with pink/purple gradient
missing_values = missing_values_table(df)
missing_values.style.background_gradient(cmap='RdPu', subset=['% of Total Values'])

### **🔍 Data Completeness Analysis:**

**Critical Findings:**
- **Xylene (61% missing)**: Severe data gap — consider excluding from analysis
- **PM10 (38% missing)**: High missing rate for a primary pollutant — requires careful handling
- **VOCs (Benzene, Toluene: 19-27% missing)**: Moderate gaps in organic compounds
- **Core pollutants (PM2.5, NO2, CO, SO2, O3)**: <20% missing — manageable with imputation
- **Essential columns (AQI, City, Date)**: 0% missing — perfect for analysis

**Strategic Approach:**
1. **Retain** PM2.5, NO2, CO, SO2, O3 (impute missing values)
2. **Evaluate** removing Xylene due to insufficient data
3. **Note** PM10 limitations in bivariate analysis
4. **Proceed** with core pollutant set for reliable modeling

---



In [ ]:
df.describe() #summary for numeric columns


PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775     21.694928     15.811136   
min        0.010000      0.000000      0.010000      0.010000      0.000000   
25%        8.580000      0.510000      5.670000     18.860000      0.120000   
50%       15.850000      0.890000      9.160000     30.840000      1.070000   
75%       30.020000      1.450000     15.220000     45.570000      3.080000   
max      352.890000    175.810000    193.860000    257.730000    455.030000   

            Toluene        Xylene           AQI  
count  21490.000000  11422.000000  24850.000000  
mean       8.700972      3.070128    166.463581  
std       19.969164      6.323247    140.696585  
min        0.000000      0.000000     13.000000  
25%        0.600000      0.140000     81.000000  
50%        2.970000      0.980000    118.000000  
75%        9.150000      3.350000    208.000000  
max      454.850000    170.370000   2049.000000

In [ ]:
# Get actual statistics
stats = df.describe()
print("📊 YOUR Dataset Statistics:")
print(f"PM2.5 mean: {stats.loc['mean', 'PM2.5']:.2f}")
print(f"PM2.5 std: {stats.loc['std', 'PM2.5']:.2f}")
print(f"AQI mean: {stats.loc['mean', 'AQI']:.2f}")
print(f"AQI std: {stats.loc['std', 'AQI']:.2f}")

📊 YOUR Dataset Statistics:
PM2.5 mean: 67.45
PM2.5 std: 64.66
AQI mean: 166.46
AQI std: 140.70


### **📊 Statistical Summary**

**Key Findings from Your Data:**

**PM2.5 (mean = 67.45 µg/m³)**: **2.7 times higher** than WHO's 24-hour safe limit (25 µg/m³), indicating **chronic excessive pollution** across Indian cities.

**AQI (mean = 166)**: Average falls in **"Unhealthy"** range (151-200) — meaning poor air quality is the **typical daily experience**, not occasional.

**High Variability**:
- PM2.5 standard deviation = 64.66 (nearly equal to the mean!) → **Extreme day-to-day swings** — some days are clean, others dangerously polluted.
- AQI standard deviation = 140.70 → Air quality fluctuates dramatically across locations and seasons.

**Range Insights**:
- Minimum AQI values suggest **some "Good" air days exist**
- Maximum values reach **hazardous levels** requiring immediate health warnings
- The wide spread confirms **both geographical and temporal pollution diversity**

**Implication**: Policy measures must address both **chronic baseline pollution** (mean) and **acute pollution events** (variability).

---



In [ ]:
df.isnull().sum() #look for missing values

City              0
Date              0
PM2.5          4598
PM10          11140
NO             3582
NO2            3585
NOx            4185
NH3           10328
CO             2059
SO2            3854
O3             4022
Benzene        5623
Toluene        8041
Xylene        18109
AQI            4681
AQI_Bucket     4681
dtype: int64

### **⚠️ Missing Values Check**

**Initial assessment reveals mixed data completeness:**

**Complete Data (0 missing):**
- City, Date → Perfect for spatial and temporal analysis

**Significant Gaps Identified:**
- **Xylene (18,109 missing)** → Severe monitoring gap (consider exclusion)
- **PM10 (11,140 missing)** → Major gap for a key pollutant
- **AQI (4,681 missing)** → Target variable has notable gaps requiring cleaning

**Moderate Gaps:**
- PM2.5, NO2, CO, SO2, O3 (2,000-4,000 missing) → Manageable with imputation

#####**Implication:** AQI missing values must be addressed first, as they directly impact our analysis objectives.
---



# **3. Data Quality Assessment & Cleaning**
---



In [ ]:
df = df.drop_duplicates()  #removes duplicate rows if any are found

### **🧹 Initial Data Cleaning**

**Duplicate Removal**: Checking for and removing exact duplicate rows to ensure each observation is unique. Duplicates could skew analysis by over-representing certain measurements.

---



In [ ]:
numeric_columns = df.select_dtypes(include=['float64']).columns  #selects only numeric columns for analysis
df[numeric_columns].isnull().mean().sort_values(ascending=False) #calculates the proportion of missing values in each numeric column


Xylene     0.613220
PM10       0.377231
NH3        0.349734
Toluene    0.272290
Benzene    0.190410
AQI        0.158511
PM2.5      0.155701
NOx        0.141715
O3         0.136196
SO2        0.130507
NO2        0.121398
NO         0.121296
CO         0.069723
dtype: float64

### **📈 Missing Data Proportion**

Calculating the **percentage of missing values** per numeric column helps prioritize handling strategies. Columns with >30% missing may require different treatment than those with <5% gaps.

---



In [ ]:
df_clean = df.copy()  #makes a separate working copy of the dataset

### **🔄 Creating a Working Copy**

A separate DataFrame `df_clean` is created to preserve the original data while applying cleaning operations. This follows best practices for reproducible data processing.

---



In [ ]:
df_clean = df_clean.dropna(subset=['AQI'])  #removes rows where AQI is missing

### **🎯 Critical Cleaning Decision**

**Rows with missing AQI are removed** because AQI is our target variable — we cannot analyze or predict what we don't measure.

**Missing pollutant values are retained** because:
1. They represent real monitoring gaps (not errors)
2. Removing them would eliminate valuable AQI observations
3. They can be handled during modeling (e.g., imputation)

This balances data quality with analysis completeness.

---



In [ ]:
df_clean.shape  #displays the new dataset size after cleaning


(24850, 16)

### **✅ Cleaning Results**

**Dataset reduced from 29531 to 24,850 rows** after removing records without AQI values. All remaining rows now have complete AQI data — essential for reliable analysis.

---



In [ ]:
df_clean.isnull().sum()  #confirms that AQI no longer contains missing values

City              0
Date              0
PM2.5           678
PM10           7086
NO              387
NO2             391
NOx            1857
NH3            6536
CO              445
SO2             605
O3              807
Benzene        3535
Toluene        5826
Xylene        15372
AQI               0
AQI_Bucket        0
dtype: int64

### **🏁 Final Data Quality Check**

Confirming that AQI column now has **zero missing values**. Remaining gaps in pollutant columns will be addressed during feature engineering or modeling stages as needed.

---

